### PR TITLE
feat(ethereum-forks): remove total difficulty for hardfork check

### LINKED ISSUE #13174: Remove extra Total difficulty argument for execution
### Describe the feature

currently we always provide block +  TTD for execution:

https://github.com/paradigmxyz/reth/blob/f82a20a616f3f1e6abdff973cb8337a8b844902a/crates/evm/execution-types/src/execute.rs#L5-L12

this is only relevant for pre-merge blocks because the hardfork is determined by the total difficulty.

this only affects sepolia+mainnet.

it would simplify a lot of things if we could only rely on the `Block` as input.
since all of these transition blocks are finalized and we sync in reverse we should be able to use the hardcoded transition blocks in the chainspec:

https://github.com/paradigmxyz/reth/blob/f82a20a616f3f1e6abdff973cb8337a8b844902a/crates/ethereum-forks/src/hardfork/ethereum.rs#L74-L93

and use that when we determine the appropriate revm specid:

https://github.com/paradigmxyz/reth/blob/f82a20a616f3f1e6abdff973cb8337a8b844902a/crates/ethereum/evm/src/config.rs#L28-L28

so that:

https://github.com/paradigmxyz/reth/blob/f82a20a616f3f1e6abdff973cb8337a8b844902a/crates/ethereum-forks/src/forkcondition.rs#L84-L87

no longer needs to check any total difficulty input

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by aroralanuk**
- fixes https://github.com/paradigmxyz/reth/issues/13174

### CODE REVIEW

**Comment by aroralanuk** [comment]
@mattsse, is this how you were thinking of removing total difficulty and relying solely on the block number (given Paris is always the merge fork)?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/ethereum-forks/src/forkcondition.rs [hunk 1]
```diff
@@ -7,18 +8,6 @@ use alloy_primitives::{BlockNumber, U256};
...
-        /// The block number at which TTD is reached, if it is known.
-        ///
-        /// This should **NOT** be set unless you want this block advertised as [EIP-2124][eip2124]
-        /// `FORK_NEXT`. This is currently only the case for Sepolia and Holesky.
-        ///
-        /// [eip2124]: https://eips.ethereum.org/EIPS/eip-2124
-        fork_block: Option<BlockNumber>,
-        /// The total difficulty after which the fork is activated.
-        total_difficulty: U256,
-    },
```
[the remark this answers is not in the document]
Comment (Rjected) [reply crates/ethereum-forks/src/forkcondition.rs:21]: yes, we need to keep one eip-2124 fork block at least. We would want _another_ blocknum field, that is used for checking if the merge is active, but not for the forkid.

**Comment by mattsse** [comment]
@aroralanuk so the way we should solve this is, we keep the `TTD` variant but add a new mandatory `activation_block_number` field, because we assume there won't be any more total difficulty hardforks and all activation numbers are known

File: crates/chainspec/src/spec.rs [hunk 2]
```diff
@@ -607,6 +607,7 @@ impl From<Genesis> for ChainSpec {
                 hardforks.push((
                     EthereumHardfork::Paris.boxed(),
                     ForkCondition::TTD {
+                        activation_block_number: genesis.config.merge_netsplit_block.expect("Merge netsplit block is required"),
```
Comment (aroralanuk) [inline crates/chainspec/src/spec.rs:610]: @mattsse @Rjected, do I need to update `merge_netsplit_block` to not be optional in the genesis file?

File: crates/consensus/common/src/calc.rs [hunk 3]
```diff
@@ -24,10 +24,10 @@ use reth_chainspec::{EthereumHardfork, Hardforks};
 pub fn base_block_reward(
     chain_spec: impl Hardforks,
     block_number: BlockNumber,
-    block_difficulty: U256,
-    total_difficulty: U256,
+    _block_difficulty: U256,
+    _total_difficulty: U256,
 ) -> Option<u128> {
-    if chain_spec.fork(EthereumHardfork::Paris).active_at_ttd(total_difficulty, block_difficulty) {
+    if chain_spec.fork(EthereumHardfork::Paris).active_at_ttd(block_number) {
```
Comment (aroralanuk) [inline crates/consensus/common/src/calc.rs:30]: should I keep the active_at_ttd function as it is and add a new active_at_merge method instead?

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 4]
```diff
@@ -393,17 +393,12 @@ where
...
-            .ok_or_else(|| BlockchainTreeError::CanonicalChain { block_hash: block.parent_hash })?;
-
-        // Pass the parent total difficulty to short-circuit unnecessary calculations.
         if !self
             .externals
             .provider_factory
             .chain_spec()
             .fork(EthereumHardfork::Paris)
-            .active_at_ttd(parent_td, U256::ZERO)
+            .active_at_ttd(block.number)
```
Comment (mattsse) [inline crates/blockchain-tree/src/blockchain_tree.rs:401]: I believe this should use the parent block number

File: crates/consensus/common/src/calc.rs (hunk 3 above)
Comment (mattsse) [reply crates/consensus/common/src/calc.rs:30]: let's keep this for now to keept changes small

**Review by mattsse — COMMENTED** [review]
cool, this looks good overall,

just need some additional time to take a closer look

**Comment by emilianobonassi** [comment]
not sure how much is related but would be great to remove ttd from the response in the block

looks like execution api ratified that and geth recently implemented (seen some slow rollout these days by rpc providers)

https://github.com/ethereum/go-ethereum/pull/30386

**Comment by aroralanuk** [comment]
thanks @Rjected

**Review by mattsse — APPROVED** [review]
amazing, this also unlocks a few nice followups

**Comment by Rjected** [comment]
note that we will have a bit of redundancy in:
* `EthereumHardfork::Paris.mainnet_activation_block()`,
* The new `activation_block` field as part of the TTD fork, and
* The `paris_block_and_final_difficulty` field on `ChainSpec`

So we can probably simplify even more things w.r.t our hardfork traits / structs

**Pull request merged**